### ISSUE TITLE
Chain of dependencies from explicit source not found

### ISSUE THREAD

**Issue by ntaylorwss** [python-poetry/poetry#8198]
- **Poetry version**: 1.5.1
- **Python version**: 3.11.4
- **OS version and name**: Arch Linux, Linux 6.1.35-1-lts

- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have consulted the [FAQ](https://python-poetry.org/docs/faq/) and [blog](https://python-poetry.org/blog/) for any relevant entries or release notes.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option) and have included the output below.

## Issue

I have a monorepo configuration where packages in `library/` depend on each other. Library packages are published to a private PyPI, currently being run through a local Docker container with the pypiserver image.

My issue comes when library package dependencies are beyond a single step. For example, if `library.a` depends on `library.b`, then when I create the lock for `library.a` with `cd library/a && poetry lock`, everything is fine. If `library.b` then depends on `library.c`, then when I try again...

```
~> cd library/a
~> poetry lock
...
  RuntimeError

  Unable to find installation candidates for library-c (0.0.0)

  at ~/.pyenv/versions/3.11.4/lib/python3.11/site-packages/poetry/installation/chooser.py:76 in choose_for
       72│ 
       73│             links.append(link)
       74│ 
       75│         if not links:
    →  76│             raise RuntimeError(f"Unable to find installation candidates for {package}")
       77│ 
       78│         # Get the best link
       79│         chosen = max(links, key=lambda link: self._sort_key(package, link))
```

The `library/a/pyproject.toml` dependency line is:
`library-b = { source = "mypypi", version = "^0.0.0" }`

The `library/b/pyproject.toml` dependency line is:
`library-c = { source = "mypypi", version = "^0.0.0" }`

In both `library/a/pyproject.toml` and `library/b/pyproject.toml`, I have:
```
[[tool.poetry.source]]
name = "mypypi"
url = "http://localhost:5050"
priority = "explicit"
```

It appears as though the `source` of "dependencies of dependencies" is not being respected? Interestingly, if I include both of the following in `library/a/pyproject.toml`:
`library-c = { source = "mypypi", version = "^0.0.0", optional = true }`

```
[[tool.poetry.extras]
myextra = ["library-c"]
```

Then `poetry lock` and `poetry install` work fine, even though `library-c` is not being installed.

In short: how do I configure a chain of dependencies from an explicit source?

**Comment by ntaylorwss** [comment]
It's amazing what making a public issue does to your problem solving ability.

After reading a bit more of the documentation on `priority = "explicit"`, I decided to change it to `"supplemental"` and include `PyPI` as a `priority = "primary"` source. That appears to have solved the problem. Hope this helps anyone similarly confused in the future.
